c-dependencies/js-compute-runtime/builtins/shared/text-decoder.cpp

```diff
@@ -17,10 +18,33 @@ bool TextDecoder::decode(JSContext *cx, unsigned argc, JS::Value *vp) {
     return false;
   }
 
-  JS::RootedString str(
-      cx, JS_NewStringCopyUTF8N(cx, JS::UTF8Chars((char *)data->data(), data->size())));
-  if (!str)
+  auto fatal =
+      JS::GetReservedSlot(self, static_cast<uint32_t>(TextDecoder::Slots::Fatal)).toBoolean();
+  auto ignoreBOM =
+      JS::GetReservedSlot(self, static_cast<uint32_t>(TextDecoder::Slots::IgnoreBOM)).toBoolean();
+
+  JS::RootedString str(cx);
+  if (ignoreBOM) {
+    auto chars = JS::UTF8Chars((char *)data->data(), data->size());
+    size_t length;
+    auto lossyUTF8Chars = JS::LossyUTF8CharsToNewTwoByteCharsZ(cx, chars, &length, js::MallocArena);
+    JS::UniqueTwoByteChars lossyChars(static_cast<char16_t *>(JS_malloc(cx, length + 1)));
```

Comment:
Is the `length + 1` necessary? You're passing the length into `JS_NewUCStringDontDeflate`, which suggests it doesn't need to be null-terminated?